astropy/stats/funcs.py

```diff
@@ -766,26 +778,59 @@ def median_absolute_deviation(a, axis=None, func=None):
         # See https://github.com/numpy/numpy/issues/7330 why using np.ma.median
         # for normal arrays should not be done (summary: np.ma.median always
         # returns an masked array even if the result should be scalar). (#4658)
-        if isinstance(a, np.ma.MaskedArray):
+        if isinstance(data, np.ma.MaskedArray):
+            is_masked = True
             func = np.ma.median
+            if ignore_nan:
+                data = np.ma.masked_where(np.isnan(data), data)
+        elif ignore_nan:
+            is_masked = False
+            if NUMPY_LT_1_10:
+                # older versions don't have nanmedian, but nanmedian
+                # is generally faster than np.ma.median (at least, that is implied
+                # in
+                # https://github.com/astropy/astropy/pull/5232#issuecomment-239173872)
+                data = np.ma.masked_invalid(data)
+                func = np.ma.median
+            else:
+                # faster when axis is None & numpy is recent
+                func = np.nanmedian
         else:
+            is_masked = False
             func = np.median
+    else:
+        is_masked = None
 
-    a = np.asanyarray(a)
-    a_median = func(a, axis=axis)
+    if not ignore_nan and NUMPY_LT_1_10 and np.any(np.isnan(data)):
+        warn("Numpy versions <1.10 will return a number rather than NaN for "
+             "the median of arrays containing NaNs.  This behavior is "
+             "unlikely to be what you expect.")
+
+    data = np.asanyarray(data)
+    data_median = func(data, axis=axis)
 
     # broadcast the median array before subtraction
```

Comment:
`np.nanmedian` has a `keepdims` argument... (it was only added in 1.10 for `np.ma.median`). But since any user function can be passed in, it is probably best to keep this (maybe add a comment, though).